### PR TITLE
feat(otel): OpenTelemetry Integration for py-shiny

### PR DESCRIPTION
**Pull request by schloerke**
## Summary

Adds comprehensive OpenTelemetry (OTel) instrumentation to py-shiny, enabling distributed tracing and observability for Shiny applications. This implementation completes all 10 phases outlined in Epic #2131, providing production-ready telemetry infrastructure with granular collection controls.

### ✅ Phase 1: Foundation (Core OTel Infrastructure) - #2142
- Core OTel infrastructure with lazy tracer/logger initialization
- Collection level management system (5 granular levels: NONE, SESSION, REACTIVE_UPDATE, REACTIVITY, ALL)
- Span wrapper utilities for async context propagation
- Environment variable configuration via `SHINY_OTEL_COLLECT`
- 22 unit tests with >90% coverage

### ✅ Phase 2: Session Lifecycle Instrumentation - #2144
- Session start/end spans wrapping the server function
- Session-ended callback spans with exception handling
- Source location attribution for server functions
- 9 comprehensive tests

### ✅ Phase 3: Reactive Flush Instrumentation - #2146
- Reactive update cycle spans (`reactive.update` spans)
- Request priority tracking and invalidation counts
- Fixed session.start span scope (refined from Phase 2)
- 7 tests covering flush cycles and metrics

### ✅ Phase 4: Reactive Execution Instrumentation - #2150
- Individual reactive execution spans for Calc, Effect, and outputs
- Namespace-aware span labeling (e.g., `"calc mod1:my_calc"`)
- Source code attribution with line numbers
- Output rendering spans with distinguishing attributes
- 49 tests including nested modules and edge cases

### ✅ Phase 5: Value Updates and Logging - #2169
- INFO-level logs for reactive value updates (`Value._set()`)
- Detailed attributes: old/new values, source location, namespace
- Respects collection level controls (requires REACTIVITY level)
- 35 tests covering various value update scenarios

### ✅ Phase 6: Error Handling and Sanitization - #2175
- Exception recording on all spans with error sanitization
- Filters out sensitive patterns (API keys, passwords, tokens)
- Preserves error type, message structure, and stack traces
- Handles chained exceptions (`raise from`) correctly
- 21 tests including security validation

### ✅ Phase 7: User-Facing API - #2178
- `otel.suppress` decorator/context manager to disable Shiny's internal telemetry for sensitive operations
- `otel.collect` decorator/context manager to re-enable telemetry when the global default has been lowered
- `otel.get_level()` to query the current effective collection level
- Clean module-attribute API: `from shiny import otel; @otel.suppress`
- Example app demonstrating collection control
- 36 tests for collection control API

### ✅ Phase 8: Testing Infrastructure and Production Readiness - #2181
- Late OTel initialization support (removed caching from `is_otel_tracing_enabled()`)
- 10 async context propagation tests (TaskGroup, as_completed, pre-created tasks)
- Download handler span instrumentation (9 tests)
- Input/clientdata filtering tests
- **Critical production readiness review** resolved 12 issues:
  - Security: Error sanitization in spans
  - Robustness: Context variable override pattern, proper cleanup
  - Performance: Conditional span creation
  - Portability: Cross-platform path handling with pathlib
  - Maintainability: Decoupled tests, clean documentation

### ✅ Phase 9: Documentation - #2184
- 700+ line module docstring in `shiny/otel/__init__.py`
- Comprehensive collection level documentation
- Quick Start guide with inline examples
- API documentation via quartodoc

### ✅ Phase 10: Follow-up Evaluation - #2184
- Documented `[otel]` optional dependency group approach
- Confirmed API-required, SDK-optional design
- Added ExtendedTask instrumentation (7 tests in reactive execution suite)
- Changed value update logs from DEBUG to INFO for better visibility
- Renamed example folder to `open-telemetry`

### Collection Levels
```python
from shiny import otel

# Control collection granularity via environment variable:
# "none" (0): No telemetry collected
# "session" (1): Session lifecycle spans only
# "reactive_update" (2): SESSION + reactive flush cycles
# "reactivity" (3): REACTIVE_UPDATE + individual reactive execution + value logs
# "all" [Default]: Currently equivalent to "reactivity"
```

### Public API
```python
from shiny import otel

# As decorator (no-parens) — stamps a single function
@otel.suppress          # Disables Shiny's internal telemetry for sensitive operations
def result_private():
    return secret_value()

@otel.collect           # Re-enables telemetry even when global default is suppressed
def result_instrumented():
    return computed_value()

# As context manager — controls telemetry during reactive object creation
with otel.suppress():
    sensitive_calc = reactive.calc(lambda: api_call_with_credentials())

with otel.collect():
    instrumented_calc = reactive.calc(lambda: compute())

# Query the current effective level
level = otel.get_level()
```

### Configuration
```bash
# Control collection level via environment variable
export SHINY_OTEL_COLLECT=reactivity  # or: none, session, reactive_update, all
python app.py
```

### Span Hierarchy
```
session.start
├── reactive.update (flush cycle 1)
│   ├── calc my_calc
│   ├── effect my_effect
│   └── output my_output
├── reactive.update (flush cycle 2)
│   ├── calc another_calc
│   └── effect another_effect
├── extended_task my_task
├── download my_download
└── session.ended
```

### Architecture
- **Lazy initialization**: Tracer/logger initialized on first use, supports late SDK configuration
- **No-op behavior**: Zero overhead when OTel SDK not configured
- **Async-aware**: Built-in context propagation via `opentelemetry.context` and Python's `contextvars`
- **Type safety**: Full pyright compliance with comprehensive type annotations
- **Error sanitization**: Prevents sensitive data leakage (API keys, passwords, tokens)
- **Cross-platform**: Robust path handling for Windows/macOS/Linux

### Integration Points
- Session lifecycle: `AppSession._run()` and session-ended callbacks
- Reactive updates: `ReactiveEnvironment.flush()`
- Reactive execution: `Context.execute_flush_callbacks()`, `Effect_._run()`, `Calc_.update_value()`
- Output rendering: `Outputs.__call__` with distinguishing attributes
- Value updates: `Value._set()` with detailed logging
- Extended tasks: `ExtendedTask` span wrapping with exception recording
- Download handlers: `session.send_progress()` download path instrumentation

### Attributes Captured
- **Source references**: File path, line number, column number, function name (via `inspect.unwrap()`)
- **Namespaces**: Module hierarchy preserved in span names
- **Reactive metadata**: Value types, invalidation counts, request priorities
- **Session info**: Session ID (explicit pattern by design)
- **Error context**: Sanitized exception types, messages, and stack traces
- **Task metadata**: Extended task queuing and execution timing

## Testing

**202 OTel unit tests** + Playwright tests:
- Foundation: 22 tests
- Session lifecycle: 9 tests
- Reactive flush: 7 tests
- Reactive execution: 49 tests (includes ExtendedTask)
- Value logging: 35 tests
- Error handling: 21 tests
- Collection control: 36 tests
- Async context: 10 tests
- Logfire integration: 4 tests
- Download spans: 9 tests

**Playwright tests:**
- Async context propagation: 4 tests (`tests/playwright/shiny/otel/async/`)
- Client data/inputs filtering: 3 tests (`tests/playwright/shiny/otel/client-data-inputs-init/`)

**Test Coverage:**
- Unit tests for all core functionality
- End-to-end Playwright tests for integration scenarios
- Edge cases: wrapped functions, deeply nested modules, async contexts
- Security: Sensitive data filtering validation
- Performance: Conditional span creation optimization
- Portability: Cross-platform path handling

### New Files
- `shiny/otel/__init__.py` - Public API (`otel.suppress`, `otel.collect`, `otel.get_level`) with comprehensive documentation
- `shiny/otel/_core.py` - Tracer/logger initialization
- `shiny/otel/_collect.py` - Collection level management
- `shiny/otel/_span_wrappers.py` - Async-aware span utilities (includes download spans)
- `shiny/otel/_attributes.py` - Source reference extraction
- `shiny/otel/_labels.py` - Namespace-aware span naming
- `shiny/otel/_errors.py` - Error sanitization
- `shiny/otel/_function_attrs.py` - Function marking helpers
- `shiny/otel/_constants.py` - OpenTelemetry semantic conventions
- `shiny/otel/_decorators.py` - User-facing decorators
- `tests/pytest/test_otel_*.py` - 10 test files (202 unit tests)
- `tests/playwright/shiny/otel/*/` - 2 Playwright test suites
- `examples/open-telemetry/` - Example app with collection control
- `_dev/claude/otel/` - Implementation plan and review documents

### Modified Files
- `shiny/session/_session.py` - Session lifecycle instrumentation, download spans
- `shiny/reactive/_core.py` - Reactive flush instrumentation
- `shiny/reactive/_reactives.py` - Value update logging, function marking
- `shiny/render/renderer/_renderer.py` - Output rendering spans
- `shiny/types.py` - Extended task instrumentation
- `pyproject.toml` - Added `opentelemetry-api>=1.20.0` dependency, `[otel]` extras group
- `docs/_quartodoc-core.yml` - Added `otel` module to API docs
- `CHANGELOG.md` - Documented all changes
- `CLAUDE.md` - Updated code style preferences

## Breaking Changes

None. This is a purely additive feature with opt-in behavior.

## Migration Guide

No migration required. OpenTelemetry support is opt-in:

### Basic Setup
```python
# 1. Install OpenTelemetry SDK (optional [otel] extras group)
# pip install "shiny[otel]"

# 2. Configure before importing shiny (or after - late init supported)
from opentelemetry import trace
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import ConsoleSpanExporter, SimpleSpanProcessor

provider = TracerProvider()
provider.add_span_processor(SimpleSpanProcessor(ConsoleSpanExporter()))
trace.set_tracer_provider(provider)

# 3. Import and use Shiny normally
from shiny import App, ui
```

### Using Logfire
```python
# Install logfire (includes opentelemetry-sdk)
# pip install logfire

# Configure before importing shiny
import logfire
logfire.configure()

# Shiny automatically detects and uses your TracerProvider
from shiny import App, ui
```

### Collection Control
```bash
# Environment variable (app-wide default)
export SHINY_OTEL_COLLECT=session
python app.py
```

```python
# Programmatic control (function/reactive-object-specific)
from shiny import otel

def server(input, output, session):
    @render.text
    @otel.suppress  # Disable telemetry for sensitive outputs
    def result_private():
        return api_call_with_credentials()

    @render.text
    @otel.collect   # Re-enable telemetry when global default is suppressed
    def result_instrumented():
        return compute()
```

## Production Readiness ✅

This implementation is **production-ready** with:
- ✅ **Security**: Error sanitization prevents sensitive data leakage
- ✅ **Performance**: Lazy initialization, no-op behavior, conditional span creation
- ✅ **Robustness**: Comprehensive error handling, proper cleanup, cross-platform support
- ✅ **Observability**: 5 collection levels for granular control
- ✅ **Testing**: 202 unit tests + Playwright tests covering unit, integration, edge cases, and security
- ✅ **Documentation**: 700+ line module docstring, API docs, examples
- ✅ **Maintainability**: Clean code, consistent patterns, decoupled tests
- ✅ **Compatibility**: Works with all OTel-compatible backends (Logfire, Jaeger, etc.)

## Related Issues

- Epic: #2131
- Phase 1: #2142 ✅
- Phase 2: #2144 ✅
- Phase 3: #2146 ✅
- Phase 4: #2150 ✅
- Phase 5: #2169 ✅
- Phase 6: #2175 ✅
- Phase 7: #2178 ✅
- Phase 8: #2181 ✅
- Phase 9 & 10: #2184 ✅
- otel.suppress/otel.collect API: #2190 ✅
- Download spans; remove OTel from bookmark callbacks: #2191 ✅

---

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: .claude/skills/gh-pr-review/SKILL.md [hunk 1]
```diff
@@ -0,0 +1,205 @@
+---
+name: gh-pr-review
```
Comment (elnelson575) [inline .claude/skills/gh-pr-review/SKILL.md]: Consolidate and/or remove given the other two below

File: .claude/skills/pr-threads-address/SKILL.md [hunk 2]
```diff
@@ -0,0 +1,74 @@
+---
+name: pr-threads-address
```
Comment (elnelson575) [inline .claude/skills/pr-threads-address/SKILL.md]: Move to posit skills repo

File: .claude/skills/pr-threads-resolve/SKILL.md [hunk 3]
```diff
@@ -0,0 +1,70 @@
+---
+name: pr-threads-resolve
```
Comment (elnelson575) [inline .claude/skills/pr-threads-resolve/SKILL.md]: Also bring over to posit skills!

File: .claude/skills/gh-pr-review/SKILL.md (hunk 1 above)
Comment (schloerke) [reply .claude/skills/gh-pr-review/SKILL.md]: https://github.com/posit-dev/skills/pull/30

File: .claude/skills/pr-threads-resolve/SKILL.md (hunk 3 above)
Comment (schloerke) [reply .claude/skills/pr-threads-resolve/SKILL.md]: https://github.com/posit-dev/skills/pull/30

**Comment by schloerke** [comment]
I'm very happy with the recent api changes. Am going to merge the PR. Any followup changes will be done in a new PR.
